x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/rules/logical/OptimizerRules.java

```diff
@@ -55,12 +58,20 @@ public OptimizerExpressionRule(TransformDirection direction) {
         @Override
         public final LogicalPlan apply(LogicalPlan plan, LogicalOptimizerContext ctx) {
             return direction == TransformDirection.DOWN
-                ? plan.transformExpressionsDown(expressionTypeToken, e -> rule(e, ctx))
-                : plan.transformExpressionsUp(expressionTypeToken, e -> rule(e, ctx));
+                ? plan.transformExpressionsDown(this::shouldVisit, expressionTypeToken, e -> rule(e, ctx))
+                : plan.transformExpressionsUp(this::shouldVisit, expressionTypeToken, e -> rule(e, ctx));
         }
 
         protected abstract Expression rule(E e, LogicalOptimizerContext ctx);
 
+        protected boolean shouldVisit(Node<?> node) {
+            return switch (node) {
+                case EsRelation esr -> false;
+                case Project p -> false;// this covers both keep and project
```

Comment:
Suggestion: IMO we should document in the javadoc that relation + projection are getting skipped per default and that one should override `shouldVisit` to change that.

Reply:
Fair point. I opened a draft at this point to see if this breaks a lot of tests (it did not :tada: ) so now I will focus no documenting and testing it.